### ISSUE TITLE
Flutter is too tied to Material Design padding

### ISSUE THREAD

**Issue by SteveAlexander** [flutter/flutter#21026]
Overall I'm enjoying using Flutter. So much so that in the media company where I work, we're switching our default app development stack from React Native to Flutter. Also, Dart is a pleasure to develop in.

Considering this, please read the following as meant respectfully, and from a place of heartfelt gratitude...

I am becoming increasingly irritated by what I see as a kind of "nanny state" enforcement of by-the-book Material Design guidelines in what are otherwise extremely useful and well thought-out Flutter components.

I have come across a number of these:

- I wanted to nudge a prefix icon a few pixels down in a search box. The icon is a magnifying-glass, and if you look at it as pseudo-typography, the handle of the magnifier should extend below the baseline of the text. I found no way to do that, while still getting the benefit of standard text decoration, because icons are constrained have a certain minimum size and border — because Material Design. So, I thought to write my own InputDecorator that would go outside the strict bounds of Material Design, but found this impossible to get a TextField to actually use. https://github.com/flutter/flutter/issues/19775

- I want to use TabBar's well thought-out logic for interacting with a TabController, and handling all sorts of cases correctly. But, I want to slightly reduce the horizontal padding around the tabs. There is no way to do this other than fork Flutter and change the global kTabLabelPadding, or copy-paste TabBar and alter the padding logic there. Or reimplement my own. The TabBar widget doesn't seem particularly open to extension.

There are more cases I don't recall specifically enough to describe in writing here. I will follow up with examples as I come across them.

What would I like to be different?

I'd like Flutter to be less prescriptive about Material Design, particularly in the use of mandatory padding. I'd like to be able to use Flutter for non-Material or close-to-Material user experiences, without taking on the complexity of writing my own basic components (by which I mean text fields, tab bars, etc.), but also without having to embarrassingly explain to my UX colleagues (who respect Material Design but do not want to follow it slavishly) that Flutter thinks it knows better than they do, and it will take a seemingly disproportionate degree of effort to make what seems like a small change — at at the code level is, if I were okay with editing the Flutter sources. But I also care about carrying technical debt, so I don't want to deviate from the mainline development of flutter.

tl;dr Flutter is awesome. But it unnecessarily forces use of Material Design padding. Please make this configurable in some way.

**Comment by zoechi** [comment]
@Hixie @HansMuller

**Comment by HansMuller** [comment]
Make the Material implementation more flexible about layouts and visuals is definitely a goal for us (and for "Material Design" itself). As a practical matter, it's easiest to address the lack of flexibility one issue at a time.

I've taken a look at #19775. Please file an issue about the default padding for tabs, we could definitely be more flexible about that.

**Comment by SteveAlexander** [comment]
Thanks @HansMuller. I'll file a separate issue for the TabBar padding.

**Comment by Hixie** [comment]
FWIW, there'll always be something that you can't do with the built-in material widgets. We consider the way that you can fork the widgets entirely to be a feature for this very reason. :-)

**Comment by zoechi** [comment]
Can we close this issue in favor of more specific individual issues?

**Comment by SteveAlexander** [comment]
that would be fine with me

**Comment by hasen6** [comment]
I would like to second the built in padding issue. I like Flutter but I seem to spend half my time trying to remove padding from every widget. Can't we just add padding and margins if we need them? Half the time the padding is so hard to find and remove since it isn't even a default padding or margin parameter. 

I don't think looking at every specific issue is going to work because there are too many widgets with some kind of spacing/padding/margin there that makes layout design very difficult.

**Comment by HansMuller** [comment]
@hasen6 I agree with you. It might be preferable to factor out component layout values, including padding, into parameters and perhaps themes, however that doesn't always scale nicely. In many cases I think it's preferable to put more emphasis on composition over configuration; in other words for components whose complex layout hews to a UI spec like material, make sure that it's easy to build something similar from parts. We've taken some small steps in this direction. For example see the docs for ListTile: https://api.flutter.dev/flutter/material/ListTile-class.html.

If there are specific components whose intrinsic  layout gets in the way it would be best to open new issues that itemize where extra flexibility is needed.

**Comment by hasen6** [comment]
If you go in that direction then you end up with apps that _look_ like Flutter apps. I think flexibility is more important. Take CSS for web design, it's 100% flexible with barely any built in padding and of  which is easily removed anyway. That way people have ultimately flexibility in their web design.

Also consistency is a problem in Flutter. Like for example with the MaterialButton you have to use the shrinkwrap option to get rid of the built in padding, whereas with other widgets it might need min width set to 0 or the padding set to zero etc. Not to mention the more obvious problem which is that they come with padding built in which is less than desirable. The icon sets' images also seem to have padding/white space around them which adds to the issue even more.

In a world with so many different devices and screen sizes, this kind of thing is more important than anything. I don't see how having zero padding by default and the option to add what you need could ever cause any kind of problem. Why would it not scale nicely? If anything it would be the opposite.

**Comment by Hixie** [comment]
If you're not trying to exactly match the material design specs, I wouldn't recommend using the material library. Creating your own or using another one from pub is definitely the approach I would recommend. Such libraries can begin from the work we have in the material library by forking that, or they can be greenfield, either approach is valid.

**Comment by lukepighetti** [comment]
There is nothing here that gives explicit padding measurements. So if you create a material widget without adjustable padding it doesn't meet material spec. https://material.io/design/layout/spacing-methods.html#touch-targets

**Comment by lukepighetti** [comment]
@zoechi Requesting reopening due to Flutter's `material` library not meeting spec on handling spacing methods.

**Comment by timsneath** [comment]
Thanks @lukepighetti for the specific comment. I think it is useful to compare with the Material spec, since we're not trying to create a forked interpretation of a design system that a group of very talented individuals have built out over many years of research and implementation learnings.

So, rather than reopen the original issue (which is a request for Flutter's Material widget set to offer more flexibility than the Material spec), I'm going to open a separate issue for the keyline feature that Luke refers to. As ever, we try and keep the Flutter issues database focused on regressions, bugs, and  "shovel-ready" new feature work, so forgive us that we frequently focus on narrowing the scope of an issue to something that we can take forward. 

Thanks for using Flutter!

**Comment by lukepighetti** [comment]
@timsneath While keylining across the entire material widget set would be a fantastic feature it would be easier to allow us to override padding in the material widgets which cover the multitude of whitespace methods that material specifies.

**Comment by Hixie** [comment]
@lukepighetti I totally hear what you're saying. It can definitely be tricky when you go outside what we implement, and that's a valid pain point that I agree people are running into.

If there are specific cases where we don't allow the flexibility that the spec allows then we should absolutely support those. Please don't hesitate to file specific bugs about specific widgets (this bug specifically was very wide-ranging in scope, which doesn't make for a good tractable GitHub issue).

There will always be areas that we don't make configurable. A ListTile doesn't have elevation or rotation. We could add those, but then it won't have color tinting. We could add that, but then it wouldn't have the ability to show a magnifier when hovered. At some point we have to draw a line between what we allow to be configured and what we don't allow to be configured. We've decided, for lack of any other clearly defined line, to draw the line at what the material spec describes.

Making a widget set for Flutter isn't that difficult, especially given that you can take the material widget set, fork it wholesale, and make it do whatever you want. We actively designed Flutter to not have just one preferred widget set. We only ship material and cupertino with Flutter itself because we feel that it makes the out-of-the-box experience more tractable for new users; we definitely don't want to imply that Flutter is tied to either.

I strongly encourage people to create new widget sets for Flutter and publish them to pub.dev.
